### PR TITLE
Port all unary float functions to structured

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56151 Port all non-float unary operators to structured (and rsqrt)
* **#56082 Port all unary float functions to structured**
* #56078 Change all unary functions stubs to use TensorIteratorBase&

The native_functions.yaml changes were done by codemod using the
following script:

```
import ruamel.yaml
from ruamel.yaml.tokens import CommentToken
from ruamel.yaml.error import CommentMark
from tools.codegen.model import *  # noqa: F403

with open("aten/src/ATen/native/native_functions.yaml", "r") as f:
    contents = f.read()

yaml = ruamel.yaml.YAML()
yaml.preserve_quotes = True
yaml.width = 1000
yaml.boolean_representation = ['False', 'True']
r = yaml.load(contents)

convert = '''\
acos
acosh
asin
asinh
atan
atanh
cos
cosh
digamma
erf
erfc
erfinv
exp
expm1
exp2
lgamma
log
log10
log1p
log2
reciprocal
sigmoid
sin
sinc
sinh
special_entr
sqrt
tan
tanh'''.split()

for e in r:
    f = NativeFunction.from_yaml(e, Location("", 0))
    if f.structured or f.structured_delegate is not None:
        continue
    n = f.func.name.name.base
    if n not in convert:
        continue
    # mutate e to make changes
    if f.func.kind() == SchemaKind.out:
        e.insert(1, 'structured', True)
        e.insert(2, 'structured_inherits', 'TensorIteratorBase')
    else:
        # TODO: The .out overload assumption is not sound in general
        e.insert(1, 'structured_delegate', f'{n}.out')

        e['dispatch'].pop('CPU', None)
        e['dispatch'].pop('CUDA', None)
        e['dispatch'].pop('CPU, CUDA', None)
        e['dispatch'].pop('CompositeExplicitAutograd', None)

        *_, last_k = e.keys()
        needs_fixup = False

        if not e['dispatch']:
            if last_k == 'dispatch':
                needs_fixup = True
            del e['dispatch']

        # Manually fix up newlines at the end, because ruamel
        # made some bad life choices about where to associate trailing
        # whitespace for nested dicts; see
        # https://stackoverflow.com/questions/42172399/modifying-yaml-using-ruamel-yaml-adds-extra-new-lines
        if needs_fixup:
            *_, last_k = e.keys()
            # post_key, pre_key, post_value, pre_value
            e.ca.items[last_k] = [None, None, CommentToken('\n\n', CommentMark(0), None), None]

with open("aten/src/ATen/native/native_functions.yaml.new", "w") as f:
    yaml.dump(r, f)
```

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27777769](https://our.internmc.facebook.com/intern/diff/D27777769)

### CODE REVIEW

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 1]
```diff
@@ -678,18 +678,6 @@ static void frexp_kernel(TensorIteratorBase& iter) {
   }                                                                                              \
   REGISTER_DISPATCH(op##_stub, &op##_kernel)
 
-  #define IMPLEMENT_COMPLEX_STRUCTURED_KERNEL(op)                                                \
```
Comment (bhosmer) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:681]: why does this have to go?

**Review by bhosmer — APPROVED** [review]

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:681]: It doesn't, it just turns out it's identical to the macro above.

Comment (bhosmer) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:681]: ah is the `iter.cast_outputs()` on 689 redundant

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -515,15 +515,15 @@
...
 
 - func: asinh_(Tensor(a!) self) -> Tensor(a!)
   variants: function, method
-  dispatch:
-    CompositeExplicitAutograd: asinh_
+  structured_delegate: asinh.out
 
 - func: asinh.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
+  structured: True
+  structured_inherits: TensorIteratorBase
```
Comment (wenleix) [inline aten/src/ATen/native/native_functions.yaml:526]: curious: what does this mean?

**Review by wenleix — APPROVED** [review]

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml:526]: https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md#handling-tensoriterator

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:681]: Hmm, actually, no, that might be a bug...

Comment (bdhirsh) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:681]: Ah yeah, that was added in this PR: https://github.com/pytorch/pytorch/pull/52277. I think it's fine to remove the old macro now that all unary ops have been ported. The old un-structured kernels were written to call `cast_outputs()` directly for some reason, which is why it was added to the macro for structured kernels

**Review by bdhirsh — APPROVED** [review]

**Pull request closed without merging**